crates/linker/src/generate_dylib.rs

```diff
@@ -0,0 +1,421 @@
+
+        for (i, out_offset) in offsets.iter_mut().enumerate().take(5) {
+            // symtab, strtab and shstrtab is ignored because they are not ALLOC
+            let in_section = match i {
+                0 => sections.hash,
+                1 => sections.gnu_hash,
+                2 => sections.dynsym,
+                3 => sections.dynstr,
+                4 => sections.dynamic,
+                _ => unreachable!(),
+            };
+
+            let offset = in_section.sh_offset(endian) as usize + extra_offset;
+            let offset = round_up_to_alignment(offset, 8);
+            *out_offset = offset;
+            writer.reserve_until(offset);
+
+            let reserved_before = writer.reserved_len();
+
+            match i {
+                0 => {
+                    writer.reserve_hash(3, hash_chain_count);
+                }
+                1 => {
+                    addresses.gnu_hash = in_section.sh_addr(endian);
+                    writer.reserve_gnu_hash(1, 1, gnu_hash_symbol_count);
+                }
+                2 => {
+                    addresses.dynsym = in_section.sh_addr(endian);
+                    writer.reserve_dynsym();
+                }
+                3 => {
+                    addresses.dynstr = in_section.sh_addr(endian);
+                    writer.reserve_dynstr();
+                }
+                4 => {
+                    addresses.dynamic = in_section.sh_addr(endian);
+                    writer.reserve_dynamic(dynamic_count);
+                }
+                _ => unreachable!(),
+            }
+
+            let reserved_after = writer.reserved_len();
+
+            extra_offset += {
+                let actual_size = reserved_after - reserved_before;
+
+                actual_size.saturating_sub(in_section.sh_size(endian) as usize)
+            };
+        }
+
+        (offsets, addresses)
+    }
+}
+
+pub const fn round_up_to_alignment(width: usize, alignment: usize) -> usize {
+    match alignment {
+        0 => width,
+        1 => width,
+        _ => {
+            if width % alignment > 0 {
+                width + alignment - (width % alignment)
+            } else {
+                width
+            }
+        }
+    }
+}
+
+fn copy_file(in_data: &[u8], custom_names: &[String]) -> Result<Vec<u8>, Box<dyn Error>> {
+    let in_elf: &elf::FileHeader64<Endianness> = elf::FileHeader64::parse(in_data)?;
+    let endian = in_elf.endian()?;
+    let in_segments = in_elf.program_headers(endian, in_data)?;
+    let in_sections = in_elf.sections(endian, in_data)?;
+    let in_syms = in_sections.symbols(endian, in_data, elf::SHT_SYMTAB)?;
+    let in_dynsyms = in_sections.symbols(endian, in_data, elf::SHT_DYNSYM)?;
+
+    let get_section = |name: &[u8]| {
+        *in_sections
+            .section_by_name(Endianness::Little, name)
+            .unwrap()
+            .1
+    };
+
+    let sections = Sections {
+        hash: get_section(b".hash" as &[_]),
```

Comment:
super nit: the as &[_] could be moved into the function instead of at every call site.

Reply:
sadly not, because `b"xxx"` creates an array of bytes, and these names have different lengths, and you run into trouble with the type of the closure

Reply:
makes sense. I guess i forgot the type of byte strings in rust.